staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go

```diff
@@ -450,42 +453,52 @@ func (c *AvailableConditionController) deleteAPIService(obj interface{}) {
 		}
 	}
 	klog.V(4).Infof("Deleting %q", castObj.Name)
-	c.enqueue(castObj)
+	if castObj.Spec.Service != nil {
+		c.rebuildAPIServiceCache()
+	}
+	c.queue.Add(castObj.Name)
 }
 
-// there aren't very many apiservices, just check them all.
-func (c *AvailableConditionController) getAPIServicesFor(obj runtime.Object) []*apiregistrationv1.APIService {
+func (c *AvailableConditionController) getAPIServicesFor(obj runtime.Object) []string {
 	metadata, err := meta.Accessor(obj)
 	if err != nil {
 		utilruntime.HandleError(err)
 		return nil
 	}
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	return c.cache[metadata.GetNamespace()][metadata.GetName()]
+}
 
-	var ret []*apiregistrationv1.APIService
+func (c *AvailableConditionController) rebuildAPIServiceCache() {
```

Comment:
this is logically dependent on services too, so rebuilding a cache here is incorrect.

Reply:
Really? It's keyed by service name and namespace, regardless of whether such a service actually exists, right?

Reply:
> this is logically dependent on services too, so rebuilding a cache on only apiservice change is incorrect.

Actually, playing out the full scenario looks like this

1. threadA sees apiservice update and starts rebuilding the map
2. threadB sees service update and issues "which apiservices depend on this service".
3. threadB checks before threadA writes the new map and gets "not watched". (this is different than before this pull because the apiservice lister always matches the change notifications)
4. threadB fails to queue interested apiservice
5. threadA finishes rebuilding and queues the apiservice that threadB skipped.


it is different, but it may not be wrong.  It's certainly worth an explanation laying out the change and defending why it is safe.

Reply:
Service/endpoint handling was already working off the cache, so this didn't seem different in character to me

Reply:
wiring an add-on indexer into informer as the cache can help us make sure the caching entries refreshed before the notification arrives. so the lister-cache and addon-cache will be always consistent

Reply:
spoke with @deads2k, a comment about why a race between service/endpoint handling and apiservice cache rebuild doesn't matter would be fine. Something like "if the service/endpoint handler wins the race against the cache rebuild, it might queue a no-longer-relevant apiservice (which will get processed an extra time, which doesn't matter), and miss a newly relevant apiservice (which will get queued by the apiservice handler)"

Reply:
> wiring an add-on indexer into informer as the cache can help us make sure the caching entries refreshed before the notification arrives. so the lister-cache and addon-cache will be always consistent

we could benchmark that, but listing from cache and building the cache key takes allocations which I'd like to avoid in the service/endpoint handler path